core/handler/convert_fields.go

```diff
@@ -176,3 +180,121 @@ func RunUnsafeCode(vm *otto.Otto, code string, timeOut time.Duration) (value ott
 	}()
 	return vm.Run(code)
 }
+
+// --------------------------------------------------------------- //
+// --------------------- Downlink messages ----------------------- //
+// --------------------------------------------------------------- //
+
+type DownlinkFunctions struct {
+	// Encoder is a JavaScript function that accepts the payload as JSON and
+	// returns an array of bytes
+	Encoder string
+}
+
+// Encoder for downlink messages. Takes a JSON string as argument and return an array of bytes
+// Encode is a method of the DownlinkFunctions struct
+func (f *DownlinkFunctions) Encode(payload map[string]interface{}) ([]byte, error) { // change interface{} to map[string]interface{}
+
+	// If there is no encoder function and if the data is not of []byte type, then we can't send it
+	//if _, ok := payload.([]byte); (f.Encoder == "") && (ok == false) {
+	if f.Encoder == "" {
+		return nil, errors.New("Encoder function not set: the data you need to be []bytes")
+	}
+
+	// Here we have either a []byte payload or a encoder for our data.
+	vm := otto.New()
+	// Here we set a string variable "payload" to the JSON payload passed in arguments
+	vm.Set("payload", payload)
+	// Value will have the returning value of the Encoder function (otto.Value)
+	value, err := RunUnsafeCode(vm, fmt.Sprintf("(%s)(payload)", f.Encoder), timeOut)
+	if err != nil {
+		return nil, errors.New("Error while running the JS function")
+	}
+
+	//not necessary --> use IsObject method
+	if value.Class() != "Array" {
+		return nil, errors.New("Encoder should return an Array.")
+	}
+
+	// Export will attempt to convert the value to a Go representation and return it via an interface{} kind
+	v, err := value.Export()
+	if err != nil {
+		return nil, fmt.Errorf("Error while encoding: %s", err)
```

Comment:
Propagate `err`. Follow conventions on errors. We're already encoding in this context so we don't have to add that to the error message